File: src/CircleAI.Commerce.Integration.Xero/XeroPrimitives.cs
// XeroPrimitives.cs — (3.3.0)
//
// Xero integration primitives — token storage, tenant tracking,
// webhook recorder. HTTP plumbing is host-supplied.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace CircleAI.Commerce.Integration.Xero;

public sealed record XeroTokens(string AccessToken, string RefreshToken, DateTimeOffset ExpiresAtUtc, string IdToken);
public sealed record XeroTenant(string TenantId, string TenantName, string TenantType);
public sealed record XeroWebhookEvent(string TenantId, string ResourceType, string ResourceId, DateTimeOffset AtUtc);

public interface IXeroBoard
{
    void StoreTokens(string userId, XeroTokens t);
    XeroTokens? GetTokens(string userId);
    bool TokensExpired(string userId, DateTimeOffset now);
    void AddTenant(string userId, XeroTenant t);
    IReadOnlyList<XeroTenant> TenantsFor(string userId);
    void RecordWebhook(XeroWebhookEvent e);
    IReadOnlyList<XeroWebhookEvent> RecentEvents(int limit = 20);
}

public sealed class InMemoryXeroBoard : IXeroBoard
{
    private readonly ConcurrentDictionary<string, XeroTokens> _tokens = new(StringComparer.Ordinal);
    private readonly ConcurrentDictionary<string, List<XeroTenant>> _tenants = new(StringComparer.Ordinal);
    private readonly List<XeroWebhookEvent> _events = new();
    private readonly object _lock = new();

    public void StoreTokens(string userId, XeroTokens t) { ArgumentNullException.ThrowIfNull(t); _tokens[userId] = t; }
    public XeroTokens? GetTokens(string userId) => _tokens.GetValueOrDefault(userId);
    public bool TokensExpired(string userId, DateTimeOffset now)
    {
        if (!_tokens.TryGetValue(userId, out var t)) return true;
        return now >= t.ExpiresAtUtc;
    }
    public void AddTenant(string userId, XeroTenant t)
    {
        ArgumentNullException.ThrowIfNull(t);
        lock (_lock)
        {
            var list = _tenants.GetOrAdd(userId, _ => new List<XeroTenant>());
            if (!list.Any(x => x.TenantId == t.TenantId)) list.Add(t);
        }
    }
    public IReadOnlyList<XeroTenant> TenantsFor(string userId)
    { lock (_lock) return _tenants.TryGetValue(userId, out var l) ? l.ToArray() : Array.Empty<XeroTenant>(); }
    public void RecordWebhook(XeroWebhookEvent e) { ArgumentNullException.ThrowIfNull(e); lock (_lock) _events.Add(e); }
    public IReadOnlyList<XeroWebhookEvent> RecentEvents(int limit = 20)
    { lock (_lock) return _events.OrderByDescending(e => e.AtUtc).Take(limit).ToArray(); }
}

File: rust/src/commerce_xero/xero_primitives.rs
//! xero_primitives.rs
//!
//! (3.3.0) Xero integration primitives — Rust port of
//! `src/CircleAI.Commerce.Integration.Xero/XeroPrimitives.cs`: token storage,
//! tenant tracking, webhook recorder. HTTP plumbing is host-supplied.
//!
//! Tokens live in a `Mutex<HashMap>` (C# `ConcurrentDictionary`); the per-user
//! tenant lists and the global event log live under a single `Mutex` (the C#
//! `object _lock`), which also guards the tenant `ConcurrentDictionary`'s inner
//! `List` mutation. `RecentEvents` reproduces
//! `OrderByDescending(e => e.AtUtc)` (stable).

use std::collections::HashMap;
use std::sync::Mutex;

use chrono::{DateTime, Utc};

/// (3.3.0) A stored OAuth token set.
///
/// Mirrors `sealed record XeroTokens(string AccessToken, string RefreshToken,
/// DateTimeOffset ExpiresAtUtc, string IdToken)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct XeroTokens {
    pub access_token: String,
    pub refresh_token: String,
    pub expires_at_utc: DateTime<Utc>,
    pub id_token: String,
}

impl XeroTokens {
    /// Constructs a token set, mirroring the positional C# record constructor.
    pub fn new(
        access_token: impl Into<String>,
        refresh_token: impl Into<String>,
        expires_at_utc: DateTime<Utc>,
        id_token: impl Into<String>,
    ) -> Self {
        Self {
            access_token: access_token.into(),
            refresh_token: refresh_token.into(),
            expires_at_utc,
            id_token: id_token.into(),
        }
    }
}

/// (3.3.0) A connected Xero tenant.
///
/// Mirrors `sealed record XeroTenant(string TenantId, string TenantName,
/// string TenantType)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct XeroTenant {
    pub tenant_id: String,
    pub tenant_name: String,
    pub tenant_type: String,
}

impl XeroTenant {
    /// Constructs a tenant, mirroring the positional C# record constructor.
    pub fn new(
        tenant_id: impl Into<String>,
        tenant_name: impl Into<String>,
        tenant_type: impl Into<String>,
    ) -> Self {
        Self {
            tenant_id: tenant_id.into(),
            tenant_name: tenant_name.into(),
            tenant_type: tenant_type.into(),
        }
    }
}

/// (3.3.0) A Xero webhook event.
///
/// Mirrors `sealed record XeroWebhookEvent(string TenantId, string ResourceType,
/// string ResourceId, DateTimeOffset AtUtc)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct XeroWebhookEvent {
    pub tenant_id: String,
    pub resource_type: String,
    pub resource_id: String,
    pub at_utc: DateTime<Utc>,
}

impl XeroWebhookEvent {
    /// Constructs an event, mirroring the positional C# record constructor.
    pub fn new(
        tenant_id: impl Into<String>,
        resource_type: impl Into<String>,
        resource_id: impl Into<String>,
        at_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            tenant_id: tenant_id.into(),
            resource_type: resource_type.into(),
            resource_id: resource_id.into(),
            at_utc,
        }
    }
}

/// (3.3.0) The Xero board contract.
///
/// Mirrors `interface IXeroBoard`.
pub trait IXeroBoard {
    /// Stores (or overwrites) a user's tokens.
    fn store_tokens(&self, user_id: &str, t: XeroTokens);
    /// Looks up a user's tokens.
    fn get_tokens(&self, user_id: &str) -> Option<XeroTokens>;
    /// Whether a user's tokens are missing or expired at `now`.
    fn tokens_expired(&self, user_id: &str, now: DateTime<Utc>) -> bool;
    /// Adds a tenant for a user (deduplicated by `tenant_id`).
    fn add_tenant(&self, user_id: &str, t: XeroTenant);
    /// Tenants connected for a user.
    fn tenants_for(&self, user_id: &str) -> Vec<XeroTenant>;
    /// Records a webhook event.
    fn record_webhook(&self, e: XeroWebhookEvent);
    /// Up to `limit` most-recent events, newest-first.
    fn recent_events(&self, limit: usize) -> Vec<XeroWebhookEvent>;
}

/// The mutable state guarded by the single lock (C# `object _lock`).
#[derive(Default)]
struct XeroState {
    tenants: HashMap<String, Vec<XeroTenant>>,
    events: Vec<XeroWebhookEvent>,
}

/// (3.3.0) In-memory [`IXeroBoard`].
pub struct InMemoryXeroBoard {
    tokens: Mutex<HashMap<String, XeroTokens>>,
    state: Mutex<XeroState>,
}

impl InMemoryXeroBoard {
    /// Creates an empty board.
    pub fn new() -> Self {
        Self {
            tokens: Mutex::new(HashMap::new()),
            state: Mutex::new(XeroState::default()),
        }
    }
}

impl Default for InMemoryXeroBoard {
    fn default() -> Self {
        Self::new()
    }
}

impl IXeroBoard for InMemoryXeroBoard {
    fn store_tokens(&self, user_id: &str, t: XeroTokens) {
        self.tokens.lock().unwrap().insert(user_id.to_string(), t);
    }

    fn get_tokens(&self, user_id: &str) -> Option<XeroTokens> {
        self.tokens.lock().unwrap().get(user_id).cloned()
    }

    fn tokens_expired(&self, user_id: &str, now: DateTime<Utc>) -> bool {
        match self.tokens.lock().unwrap().get(user_id) {
            None => true,
            Some(t) => now >= t.expires_at_utc,
        }
    }

    fn add_tenant(&self, user_id: &str, t: XeroTenant) {
        let mut state = self.state.lock().unwrap();
        let list = state.tenants.entry(user_id.to_string()).or_default();
        if !list.iter().any(|x| x.tenant_id == t.tenant_id) {
            list.push(t);
        }
    }

    fn tenants_for(&self, user_id: &str) -> Vec<XeroTenant> {
        self.state
            .lock()
            .unwrap()
            .tenants
            .get(user_id)
            .cloned()
            .unwrap_or_default()
    }

    fn record_webhook(&self, e: XeroWebhookEvent) {
        self.state.lock().unwrap().events.push(e);
    }

    fn recent_events(&self, limit: usize) -> Vec<XeroWebhookEvent> {
        let mut out: Vec<XeroWebhookEvent> = self.state.lock().unwrap().events.clone();
        out.sort_by(|a, b| b.at_utc.cmp(&a.at_utc));
        out.truncate(limit);
        out
    }
}

/// The default `recent_events` limit in the C# `RecentEvents(int limit = 20)`.
pub const DEFAULT_EVENT_LIMIT: usize = 20;
